xdsl/irdl.py

```diff
@@ -335,6 +337,11 @@ class AttrSizedResultSegments(IRDLOption):
     """Name of the attribute containing the variadic result sizes."""
 
 
+_R = TypeVar("_R", bound=Attribute | type[Attribute] | AttrConstraint)
+
+S_ResultDef: TypeAlias = Annotated[_R, IRDLAnnotations.ResultDefAnnot]
```

Comment:
```suggestion
S_ResultDef: TypeAlias = Annotated[SSAValue, _R, IRDLAnnotations.ResultDefAnnot]
```
The reason for this is that during a transformation, we might change the invariants of the operands/results. Thus, we can't know statically what will be the type of an operand/result.